analyzers/src/SonarAnalyzer.CFG/ShimLayer/IOperationWrapperSonar.cs

```diff
@@ -24,37 +24,31 @@
 namespace StyleCop.Analyzers.Lightup
 {
     // This is a temporary substitute for IOperationWrapper in case StyleCop will accept PR https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3381
-    public class IOperationWrapperSonar
+    public readonly struct IOperationWrapperSonar
     {
-        private static readonly PropertyInfo ParentProperty;
-        private static readonly PropertyInfo ChildrenProperty;
-        private static readonly PropertyInfo LanguageProperty;
-        private static readonly PropertyInfo IsImplicitProperty;
-        private static readonly PropertyInfo SemanticModelProperty;
-
-        private IOperation parent;
-        private IEnumerable<IOperation> children;
-        private string language;
-        private bool? isImplicit;
-        private SemanticModel semanticModel;
-
-        public IOperation Instance { get; }
-        public IOperation Parent => ParentProperty.ReadCached(Instance, ref parent);
-        public IEnumerable<IOperation> Children => ChildrenProperty.ReadCached(Instance, ref children);
-        public string Language => LanguageProperty.ReadCached(Instance, ref language);
-        public bool IsImplicit => IsImplicitProperty.ReadCached(Instance, ref isImplicit);
-        public SemanticModel SemanticModel => SemanticModelProperty.ReadCached(Instance, ref semanticModel);
+        private static readonly Func<IOperation, IOperation> ParentAccessor;
+        private static readonly Func<IOperation, IEnumerable<IOperation>> ChildrenAccessor;
+        private static readonly Func<IOperation, string> LanguageAccessor;
+        private static readonly Func<IOperation, bool> IsImplicitAccessor;
+        private static readonly Func<IOperation, SemanticModel> SemanticModelAccessor;
 
         static IOperationWrapperSonar()
         {
             var type = typeof(IOperation);
-            ParentProperty = type.GetProperty(nameof(Parent));
-            ChildrenProperty = type.GetProperty(nameof(Children));
-            LanguageProperty = type.GetProperty(nameof(Language));
-            IsImplicitProperty = type.GetProperty(nameof(IsImplicit));
-            SemanticModelProperty = type.GetProperty(nameof(SemanticModel));
+            ParentAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<IOperation, IOperation>(type, nameof(Parent));
+            ChildrenAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<IOperation, IEnumerable<IOperation>>(type, nameof(Children));
+            LanguageAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<IOperation, string>(type, nameof(Language));
+            IsImplicitAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<IOperation, bool>(type, nameof(IsImplicit));
+            SemanticModelAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<IOperation, SemanticModel>(type, nameof(SemanticModel));
         }
 
+        public IOperation Instance { get; }
+        public IOperation Parent => ParentAccessor(Instance);
```

Comment:
sorry to ask silly questions, but given that I am the only volunteer reviewer, I need to understand now how the whole ShimLayer is working.

Why does having a `Func` reduce memory allocations? Wouldn't each property call create a new instance in memory? As opposed to using something that is cached? 

Or is rather the fact that it does create an instance just for the scope of its usage, and previously the `ReadCached` was actually storing ALL instances, even when they were not necessary?

Do I understand well that this is a tradeoff of memory usage (drop the cache) vs cycle usage (we will recreated the same `Parent` instance multiple times for the same syntax node?

Reply:
in your perf evaluation (https://github.com/SonarSource/sonar-dotnet/pull/8105#issue-1917369901) I see you only mention memory and not CPU usage difference (or I am reading it wrong?)

Reply:
Before the `IOperationWrapperSonar` was a class and instantiated via the the `IOperationWrapperSonar ToSonar(this IOperation)` extension method. The class was used to access properties like `IOperation.Parent` which was introduced in a later version of Roslyn that we reference. The old solution access the properties via reflection like so:
* A static `PropertyInfo` of the `Parent` property was created in the static constructor and stored in a static field
* For each instance of `IOperationWrapperSonar` a reflection-based call via `PropertyInfo.GetValue` was made. This is very slow. The result was therefore cached in a field of the instance.

This resulted in

* An allocation of an instance of the wrapper on each call to `ToSonar` resulting in 160MB of total allocations (the object contains 5 fields (all object references) so one instance might be roughly 5 times 8 byte + overhead). So there are a lot of instances around.
* A slow call for the first time the property __of the instance__ is called

The "normal" way the wrapper work is like so:
* Create a `readonly struct` that wraps the `IOperation` in a single field (called `Instance`). This does not allocate on the heap at all (we do not box the wrappers anywhere). It doesn't add any overhead at all as the single field is exactly as big as the original value, so the wrapper takes as much stack space as without.
* The access to the properties is done via a `Func<IOperation, TResult>` delegate. This delegate is stored in a static field and created in the static constructor. The delegate implementation is dynamically created via `System.Linq.Expressions` and compiled (via [`LambdaExpression.Compile`](https://learn.microsoft.com/dotnet/api/system.linq.expressions.lambdaexpression.compile?view=net-7.0) which generates IL code via Reflection.Emit in a dynamic generated assembly). This takes a significant time but is paid only once at the very first access of the first instance (first call to `ToSonar()`. Every subsequent call for any instance is as fast as IL generated by the compiler in a normal assembly. The generated implementation for that delegate is just a simple property access and looks about like so `(IOperation operation) => operation.Parent;`

I did not measure the performance impact on time. I don't think it is measurable in a real-world scanner run but I could give it a try in the analyzerrunner with the SymbolicExecutionRunner analyzer only.

Reply:
With the analyzer runner and SE only for CSVHelper I get the following results:

### Allocations

18MB of 1.5 GB (1%)
![image](https://github.com/SonarSource/sonar-dotnet/assets/103252490/39ac1cfb-afc9-4e28-b235-5432c2d76c89)

After: no allocations for `IOperationWrapperSonar`

### Time savings

Time spent in `PropertyInfo.GetValue` before 685 ms (0.05% of overall runtime) down to after 15ms after.

Before
Place 25th of most expensive methods
![image](https://github.com/SonarSource/sonar-dotnet/assets/103252490/900cdec4-6404-41bc-83f3-dc844303231f)


![image](https://github.com/SonarSource/sonar-dotnet/assets/103252490/d6c5e870-1c3d-4ff9-a2f5-d67c82957a4f)

After
![image](https://github.com/SonarSource/sonar-dotnet/assets/103252490/6c95f19e-9a46-4a4c-8d5a-d75567ac39a9)